docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+
+### Implementation Question
+
+* What is the format of a snapshot
+    * Complete snapshot 
+    * Ordered IAVL key ranges
+    * Compressed individually chunks which can be validated
+* How is data validated
+    * Trust a peer with it's data blindly
+    * Trust a majority of peers
+    * Use light client validation to validate any key based on consensus
+      produced merkle tree root
+* What are the performance characteristics
+    * Random vs sequential reads
+    * How parallelizeable is the scheduling algorithm
+
+### Proposals
+Broadly speaking there are two approaches to this problem which have had
+varying degrees of discussion and progress. These approach can be
+summarized as:
+
+* Lazy: Where snapshots are produced dynamically at request time
+* Eager: Where snapshots are produced periodically and served from disk at request time
+
+#### Lazy StateSync
+An [initial specification](https://docs.google.com/document/d/15MFsQtNA0MGBv7F096FFWRDzQ1vR6_dics5Y49vF8JU/edit?ts=5a0f3629) was published by Alexis Sellier.
+In this design, the state has a given `size` of primitive elements (like
+keys or nodes), each element is assigned a number from 0 to `size-1`,
+and chunks consists of a range of such elements.  Ackratos raised
+[some concerns](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit)
+about this design, somewhat specific to the IAVL tree, and mainly concerning
+performance of random reads and of iterating through the tree to determine element numbers
+(ie. elements aren't indexed by the element number).
+
+This design could be generalized to have numbered chunks instead of
+elements, where chunks are indexed by their chunk number and determined
+by some algorithm for grouping the elements in the tree. This would
+allow pre-computation and persistence of the chunks that could alleviate
+the concerns about real-time iteration and random reads.  However, there
+seems to be a larger issue with the "numbered chunks" approach: it does
+not seem possible, in general, to verify that the chunk corresponds to
+the chunk number, even if it can be verified that the chunk is a valid
+set of nodes in the tree. A malicious peer could provide a chunk that
+contains valid nodes from the tree but does not actually match the
+alleged chunk number. This could make it difficult for nodes to figure
+out which parts of the tree they have and which parts they still need.
+Hence, to go with this chunking approach to state sync, information
+about how state is chunked would need to be hashed into the state
+itself, or into the block header, so that chunk numbers could be
+directly verified.
+
+An alternative design was suggested by Jae Kwon in
+[#3639](https://github.com/tendermint/tendermint/issues/3639) where chunking
+happens lazily and in a dynamic way: nodes request key ranges from their peers,
+and peers respond with some subset of the
+requested range and with notes on how to request the rest in parallel from other
```

Comment:
This part of this design is super tricky to implement correctly in my view as we ask set of peers and each of them come back with notes how to download rest.